celery_exporter/celery_state.py

```diff
@@ -0,0 +1,169 @@
+            "received": TaskState.RECEIVED,
+            "started": TaskState.STARTED,
+            "failed": TaskState.FAILURE,
+            "retried": TaskState.RETRY,
+            "succeeded": TaskState.SUCCESS,
+            "revoked": TaskState.REVOKED,
+            "rejected": TaskState.REJECTED,
+        }
+
+        return MAPPING.get(event_kind, TaskState.UNDEFINED)
+
+
+@dataclass
+class Task:
+    uuid: str = CELERY_MISSING_DATA
+    name: str = CELERY_MISSING_DATA
+    local_received: float = 0.0
+    runtime: Optional[float] = None
+    state: TaskState = TaskState.UNDEFINED
+
+    def update_from_event(self, event: Dict):
+        kind: Optional[str] = event.get("type")
+        if kind is None:
+            raise ValueError("Invalid Event: missing type")
+        elif not (isinstance(kind, str)):
+            raise TypeError(f"Except type str for kind, found {type(kind)}")
+
+        splitted = kind.split("-")
+        state = splitted[1]
+
+        if is_task_event(kind):
+            self.uuid = event.get("uuid", CELERY_MISSING_DATA)
+            self.state = TaskState.from_event(state)
+            local_received = event.get("local_received")
+            if self.local_received is None:
+                raise ValueError("Invalid Event: missing local_received")
+            elif not (isinstance(local_received, float)):
+                raise TypeError(
+                    f"Expected type float for local_received, found {type(local_received)}"
+                )
+            self.local_received = local_received
+
+            self.name = event.get("name", CELERY_MISSING_DATA)
+            self.runtime = event.get("runtime")
+
+
+class CeleryState:
+    def __init__(
+        self,
+        event_count: int,
+        task_count: int,
+        queue_by_task: Dict[str, str],
+        tasks: LRUCache[str, Task],
+    ):
+        self.event_count = event_count
+        self.task_count = task_count
+        self.queue_by_task = queue_by_task
+        self.tasks = tasks
+
+    @classmethod
+    def new(cls, max_tasks_in_memory: int) -> "CeleryState":
+        return cls(
+            event_count=0,
+            task_count=0,
+            queue_by_task=defaultdict(),
+            tasks=LRUCache(maxsize=max_tasks_in_memory),
+        )
+
+    def event(self, task: Task):
+        self.event_count += 1
+
+        if self.tasks.get(task.uuid) is None:
+            self.tasks[task.uuid] = task
+
+        if task.state == TaskState.RECEIVED:
+            self.task_count += 1
+
+    def collect(self, event: Dict) -> CollectOutcome:
+        kind: Optional[str] = event.get("type")
+        if kind is None:
+            raise ValueError("Invalid Event: missing type")
+
+        if not (is_task_event(kind)):
+            return (None, None, None, None)
+
+        task = Task()
+        task.update_from_event(event)
+
+        name: str
+        queue: str
+        if task.state in [TaskState.SUCCESS, TaskState.FAILURE, TaskState.REVOKED]:
```

Comment:
This is kinda personal preference but I think it might be slightly faster? 🤷 

```suggestion
        if task.state in (TaskState.SUCCESS, TaskState.FAILURE, TaskState.REVOKED):
```

Reply:
Hmm very close, but no cigar 🤷 
```
In [10]: %timeit state in (TaskState.SUCCESS, TaskState.FAILURE, TaskState.REVOKED)
207 ns ± 1.47 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [11]: %timeit state in [TaskState.SUCCESS, TaskState.FAILURE, TaskState.REVOKED]
206 ns ± 1.17 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Reply:
Hah nice!